cores/esp32/esp32-hal-ledc.c

```diff
@@ -120,11 +132,11 @@ bool ledcAttachChannel(uint8_t pin, uint32_t freq, uint8_t resolution, uint8_t c
 bool ledcAttach(uint8_t pin, uint32_t freq, uint8_t resolution)
 {
     uint8_t free_channel = ~ledc_handle.used_channels & (ledc_handle.used_channels+1);
-    if (free_channel == 0 || resolution > LEDC_MAX_BIT_WIDTH){
-        log_e("No more LEDC channels available! (maximum %u) or bit width too big (maximum %u)", LEDC_CHANNELS, LEDC_MAX_BIT_WIDTH);
+    if (free_channel == 0){
+        log_e("No more LEDC channels available! (maximum is %u channels)", LEDC_CHANNELS);
         return false;
     }
-    int channel = log2(free_channel & -free_channel);
+    uint8_t channel = log2(free_channel & -free_channel);
```

Comment:
(I think) I understand the logic behind this, but isn't it a bit too much to use such a heavy function to find a specific bit not yet set to 0?
After all this log function does convert to double types and if this is the only time you use this log function it will add quite a bit of code.
And given the hardly readable bit operations which look really (a bit too) smart and fast, using this log2 function will get rid of all the gained speed and small code.

Reply:
Hmm I didn't know the CTZ. Not sure how standard C++ this is, but it will probably create the smallest code.
The while loop is the most readable one.

Not 100% sure if there should be an extra check to make sure we're not returning any channel higher than the max? (are there ESP's with less than 8 channels?)

Reply:
I will run some more test to check if the there is no risk of getting channel higher than max. But there is already a check at the top of `ledcAttachChannel()` which is called by ledcAttach:
```
if(channel >= LEDC_CHANNELS || ledc_handle.used_channels & (1UL << channel)){
        log_e("Channel %u is not available (maximum %u) or already used!", channel, LEDC_CHANNELS);
        return false;
}
```

I think as the `__buildin_ctz()` is standard and useful function in GCC, it can be used. Also can add a simple comment what it does for better readability.

Reply:
I read the `__buildin_ctz(0)` is undefined, but since you already checked for it, that's fine here.
Including some comment, it is probably the most clean code indeed.

Reply:
Call __buildin_ctz(0) will never happen, as before that there is a check for `free_channel`:
```
if (free_channel == 0){
        log_e("No more LEDC channels available! (maximum is %u channels)", LEDC_CHANNELS);
        return false;
}
```
So we have winner here :)

Reply:
Not only do we have a winner, I also learned about a new function in C/C++ :)